lib/ModelMixins/TileErrorHandlerMixin.ts

```diff
@@ -0,0 +1,369 @@
+import TileProviderError from "terriajs-cesium/Source/Core/TileProviderError";
+import ImageryProvider from "terriajs-cesium/Source/Scene/ImageryProvider";
+import when from "terriajs-cesium/Source/ThirdParty/when";
+import Constructor from "../Core/Constructor";
+import TerriaError from "../Core/TerriaError";
+import getUrlForImageryTile from "../Map/getUrlForImageryTile";
+import CommonStrata from "../Models/CommonStrata";
+import CompositeCatalogItem from "../Models/CompositeCatalogItem";
+import Model from "../Models/Model";
+import CatalogMemberTraits from "../Traits/CatalogMemberTraits";
+import MappableTraits, { RectangleTraits } from "../Traits/MappableTraits";
+import RasterLayerTraits from "../Traits/RasterLayerTraits";
+import DiscretelyTimeVaryingMixin from "./DiscretelyTimeVaryingMixin";
+
+type ModelType = Model<
+  MappableTraits & RasterLayerTraits & CatalogMemberTraits
+>;
+
+/**
+ * A mixin for handling tile errors in raster layers
+ *
+ */
+function TileErrorHandlerMixin<T extends Constructor<ModelType>>(Base: T) {
+  abstract class Klass extends Base {
+    tileFailures = 0;
+    private readonly tileRetriesByMap: Map<string, number> = new Map();
+
+    tileRetryOptions: retry.OperationOptions = {
+      retries: 8,
+      minTimeout: 200,
+      randomize: true
+    };
+
+    /**
+     * A hook that may be implemented by catalog items for custom handling of tile errors.
+     *
+     * @param maybeError A tile request promise that that fails with the tile error
+     * @param tile       The tile to be fetched
+     * @returns          A modified promise
+     *
+     * The item can then decide to retry the tile request after adding custom parameters
+     * like an authentication token.
+     */
+    handleTileError?: (
+      request: Promise<void>,
+      tile: { x: number; y: number; level: number }
+    ) => Promise<void>;
+
+    get hasTileErrorHandlerMixin() {
+      return true;
+    }
+
+    /*
+     * Handling tile errors is really complicated because:
+     *
+     * 1) things go wrong for a variety of weird reasons including server
+     * misconfiguration, servers that are flakey but not totally broken,
+     * etc.
+     *
+     * 2) we want to fail as gracefully as possible, and give flakey servers
+     * every chance chance to shine
+     *
+     * 3) we don't generally have enough information the first time something fails.
+     *
+     * There are several mechanisms in play here:
+     *   - Cesium's Resource automatically keeps trying to load any resource
+     *     that fails until told to stop, but tells us each time.
+     *   - The "retry" library knows how to pace the retries, and when to actually stop.
+     */
+    onTileLoadError(tileProviderError: TileProviderError): void {
+      const operation = retry.operation(this.tileRetryOptions);
+      // result.reject = stop trying to load this tile
+      // result.resolve = retry loading this tile
+      const result = when.defer();
```

Comment:
Yep, agree on not using `when`. Here we are forced to use it because our Cesium fork expects a `when` object here - https://github.com/TerriaJS/cesium/blob/terriajs/Source/Core/TileProviderError.js#L161.

I'll add a comment explaining that bit.

Reply:
This is done.